lib/pdfkit/source.rb

```diff
@@ -42,11 +42,45 @@ def to_s
     private
 
     def shell_safe_url
-      url_needs_escaping? ? URI::DEFAULT_PARSER.escape(@source) : @source
+      url = url_needs_escaping? ? URI::DEFAULT_PARSER.escape(@source) : @source
+
+      URI::DEFAULT_PARSER.parse(url)
+      shellescape_query(url)
     end
 
     def url_needs_escaping?
       URI::DEFAULT_PARSER.escape(URI::DEFAULT_PARSER.unescape(@source)) != @source
     end
+
+    def needs_shell_escaping?(data)
+      Shellwords.escape(shellwords_unescape(data)) != data
+    end
```

Comment:
I wouldn't bother checking this, shouldn't hurt to always escape. People should not be passing in shell formatted content to the library anyway.

Reply:
There's some specs that test whether a previously escaped source URL does not get double-escaped. Example, what if someone passes in a URL that was formatted by `URI::HTTP.build` and already contains '%XX' escaped characters? Should we assume the user must always pass in an unescaped URL? Or should we assume it's the user's responsibility to properly format the URL? Should those tests be removed?